### ISSUE TITLE
Expose pos_to_line to consumer code

### ISSUE THREAD

**Issue by Phrohdoh** [kevinmehall/rust-peg#146]
https://github.com/kevinmehall/rust-peg/blob/8098fab1515dc38ca25cdebda9b17655aa5554e1/src/translate.rs#L337

It'd be great to be able to get the line and column of tokens (my AST diagnostics makes heavy use of this).

Is this already exposed in some way?

`#position` isn't quite what I'm looking as that is just the (byte?) offset into the source string.

This is the only thing I have seen so far that is preventing me from moving from pest to rust-peg.
pest exposes this via [`input().line_col()`](http://dragostis.github.io/pest/pest/trait.Input.html#tymethod.line_col).

#85 is possibly related.

Example grammar:

```rust
peg! tsql_grammar(r#"
use ast;

whitespace0 = [ \t\r\n]*
whitespace1 = [ \t\r\n]+

#[pub]
top -> ast::TopStatement
    = kw_top:$("top"i)
      whitespace1
      paren_open:tok_paren_open?
      whitespace0
      lit:lit_i32
      whitespace0
      paren_close:tok_paren_close? {
          if (paren_open.is_some() && paren_close.is_none()) ||
             (paren_open.is_none() && paren_close.is_some()) {
                 return Err(ParseError {
                     // need to be able to create these fields
                 });
          }

          ast::TopStatement {
              top_keyword_pos: ast::Position::from((1, 1)),
              expr: ast::Node {
                  pos: ast::Position::from((1, 1)), // also need position (line, col) here
                  value: ast::Expression::Literal { lit: ast::Literal::Int(lit) },
              },

              paren_open: paren_open,
              paren_close: paren_close,
          }
      }"#);
```

**Comment by kevinmehall** [comment]
Agreed that we need a good way to do source mapping, and line numbering is the minimal subset of that. I'm reluctant to just make that method `pub` because it's injected into every generated grammar module, and I don't necessarily want those modules to have any public API besides the parse functions themselves. Functions like this are in the grammar modules to avoid the need for any extra crate to be linked into the program, not because they belong there.

I plan to add a codemap inspired by the one in rustc as a separate optional crate. As a stopgap, would it work to copy that function into your program? It's a generic function to map a byte offset to a line number in any string; doesn't depend on any PEG implementation details.

**Comment by Phrohdoh** [comment]
Yeah temporarily copying it into my code should be fine.
Just a single question for that; where do I get the full input `&str` to pass the `pos_to_line`?

**Comment by kevinmehall** [comment]
Oh, I didn't look at your example closely enough and assumed that you wanted to map these outside of the parser function, where you have access to the same string you parsed. Inside of the parser function there's a variable `__input` containing the string passed to the parser function, but that's considered an implementation detail.

It will be somewhat costly to run `pos_to_line` as I implemented it for every node in your AST. It iterates through the entire input counting lines so that would approach quadratic time. Right now it's used only once, and only in an error case.

There are certainly data structures that could be used to calculate a large number of line/column locations more efficiently. Perhaps you could create a Vec of the byte positions of newlines before parsing and pass it in using thread-local-storage (see #84). Then write a function that takes a byte position, looks up the newline positions Vec in TLS, and binary-searches for the line number in that Vec.

**Comment by Phrohdoh** [comment]
Hmm that is a pretty large trade-off then, one I don't think I can make.
How difficult do you think it'd be (if this lines up with the rest of the design of rust-peg) to have rules return/create something like https://phrohdoh.github.io/tsqlust/pest/struct.Token.html for matched slices? (basically just `{ start: usize, end: usize }`)

**Comment by Phrohdoh** [comment]
Is rust-peg's pos_to_line more or less expensive than https://github.com/dragostis/pest/blob/bb51a98f4d8df35e54495a3596cb88e78108be3d/src/inputs/string_input.rs#L78?

**Comment by kevinmehall** [comment]
> How difficult do you think it'd be (if this lines up with the rest of the design of rust-peg) to have rules return/create something like https://phrohdoh.github.io/tsqlust/pest/struct.Token.html for matched slices? (basically just { start: usize, end: usize })

That's pretty much #85 if I understand your use case correctly.

> Is rust-peg's pos_to_line more or less expensive than https://github.com/dragostis/pest/blob/bb51a98f4d8df35e54495a3596cb88e78108be3d/src/inputs/string_input.rs#L78?

They both linearly scan through the string each time they are called, so approximately the same.

**Comment by kevinmehall** [comment]
Closing this as a duplicate of #85. I think the `pos_to_line` function itself is not the right abstraction to expose -- for one, the concept of a "line" (or lack thereof) is defined by the grammar itself. It's only there to get semi-usable error messages without the need to write additional code.

**Comment by mcandre** [comment]
The definition of a source code line is fairly stable across grammars. It's the same value as your text editor shows.

The main quirk is to treat the following as equally valid delimiters: CR, LF, and CRLF.
